include/deal.II/base/template_constraints.h

```diff
@@ -41,26 +57,35 @@ template <class Base, class... Derived>
 struct is_base_of_all
 {
   static constexpr bool value =
-    std::is_same<BoolStorage<std::is_base_of<Base,Derived>::value..., true>,
-    BoolStorage<true, std::is_base_of<Base,Derived>::value...>>::value;
+    internal::TemplateConstraints::all_true<std::is_base_of<Base,Derived>::value...>::value;
 };
 
 
 
 /**
  * This struct is a generalization of std::is_same to template
- * parameter packs and tests if all of the Types... classes are
- * in Type classes. The result is stored in the member variable value.
+ * parameter packs and tests if all of the types in the `Types...`
+ * parameter pack are equal to the `Type` given as first template
+ * argument. The result is stored in the member variable value.
  */
 template <class Type, class... Types>
 struct all_same_as
 {
   static constexpr bool value =
-    std::is_same<BoolStorage<std::is_same<Type, Types>::value..., true>,
-    BoolStorage<true, std::is_same<Type, Types>::value...>>::value;
+    internal::TemplateConstraints::all_true<std::is_same<Type, Types>::value...>::value;
 };
 
 
+/*
+ * A generalization of `std::enable` that only works if
+ * <i>all</i> of the given boolean template parameters are
+ * true.
+ */
+template <bool... Values>
+struct enable_if_all : std::enable_if<internal::TemplateConstraints::all_true<Values...>::value>
```

Comment:
For this to be analogous to `std::enable_if` I would expect to be able to specify a type (default `void`) I can access via `std::enable_if_all<...>::type` if all of the statements are true. Can you add this?

Reply:
Well, that's sort of a problem. I had tried that, but to be analogous to `std::enable_if`, I would have to declare this as
```
template <bool... Values,
                 typename T = void>
struct enable_if_all : std::enable_if<..., T> {};
```
But that's not allowed: parameter packs must appear *at the end of the template argument list*. So my only choice would have been to move `T` to be the first argument, and then I can't give it a default value.

So, I didn't know what to do and went with the easy path which was to omit `T` altogether. I couldn't come up with a better solution. What do you think the best approach is, @masterleinad ?